File: go/fixture/go-mini/internal/handlers/trace.go
// Package handlers holds the HTTP handlers of the fleet service.
package handlers

import (
	"encoding/json"
	"net/http"

	"example.com/go-mini/internal/repository"
	"example.com/go-mini/internal/services"
)

// Handler is a handler.
type Handler struct {
	store repository.Store
}

// NewHandler creates a new Handler.
func NewHandler(store repository.Store) *Handler {
	return &Handler{store: store}
}

// List lists the devices as JSON.
func (h *Handler) List(w http.ResponseWriter, r *http.Request) {
	h.trace(w, r.Header.Get("X-Trace"))
	devices, err := h.store.List(r.Context())
	if err != nil {
		http.Error(w, "store unavailable", http.StatusServiceUnavailable)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(devices)
}

// trace echoes the caller's trace id so a device's log lines and the request
// that produced them can be joined later.
func (h *Handler) trace(w http.ResponseWriter, traceID string) {
	w.Header().Set("X-Trace", traceID)
}

// Routes registers the handlers on mux.
func Routes(mux *http.ServeMux, store repository.Store, svc *services.DeviceService) {
	h := NewHandler(store)
	mux.HandleFunc("GET /status", Status(store))
	mux.HandleFunc("GET /devices", h.List)
	mux.HandleFunc("POST /devices", Register(store))
	mux.HandleFunc("POST /heartbeat", Heartbeat(svc))
}

File: py/fixture/py-mini/internal/handlers/trace.py
"""The device list handler and the route table."""

from http import HTTPStatus
from typing import Any

from internal.handlers.authz import require_write
from internal.handlers.devices import register
from internal.handlers.heartbeat import heartbeat
from internal.handlers.mux import Mux, Request, Response, text, write_json
from internal.handlers.status import status
from internal.models.device import Device
from internal.models.grants import Grants
from internal.repository.store import Store
from internal.services.device_service import DeviceService


class Handler:
    """Handler is a handler."""

    def __init__(self, store: Store) -> None:
        """Create a new Handler."""
        self.store = store

    def list_devices(self, req: Request) -> Response:
        """List the devices as JSON."""
        try:
            devices = self.store.list_devices()
        except OSError:
            return text(HTTPStatus.SERVICE_UNAVAILABLE, "store unavailable")
        resp = write_json(HTTPStatus.OK, [_device_json(d) for d in devices])
        self._trace(resp, req.headers.get("x-trace", ""))
        return resp

    def _trace(self, resp: Response, trace_id: str) -> None:
        """Echo the caller's trace id.

        A device's log lines and the request that produced them can be joined later.
        """
        resp.set_header("X-Trace", trace_id)


def _device_json(d: Device) -> dict[str, Any]:
    return {
        "id": d.id,
        "tenant": d.tenant,
        "status": d.status,
        "version": d.version,
        "tags": d.tags,
        "last_seen": d.last_seen.isoformat(),
    }


def routes(mux: Mux, store: Store, svc: DeviceService, grants: Grants) -> None:
    """Register the handlers on mux; the mutating routes run only when grants allow writes."""
    h = Handler(store)
    mux.handle("GET", "/status", status(store))
    mux.handle("GET", "/devices", h.list_devices)
    mux.handle("POST", "/devices", require_write(grants, register(store)))
    mux.handle("POST", "/heartbeat", require_write(grants, heartbeat(svc)))
